lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Solve.hs

```diff
@@ -0,0 +1,131 @@
+  Class (csupers),
+  Constraint (C),
+  Rule ((:<=)),
+  mapPat,
+  ruleHeadClass,
+  ruleHeadPat,
+ )
+
+import Control.Monad.Except (throwError)
+import Control.Monad.Reader (ReaderT, runReaderT)
+import Control.Monad.Reader.Class (MonadReader (ask))
+import Control.Monad.Writer.Class (MonadWriter (tell))
+import Control.Monad.Writer.Strict (WriterT, execWriterT)
+import Data.Foldable (traverse_)
+import Data.List (foldl')
+import Data.Set qualified as S
+import Data.Text (Text)
+
+{- Pattern/Template/Unification variable  substitution.
+   Given a string that represents a variable name,
+   and a type to instantiate variables with that name to,
+   performs the instantiation
+-}
+subV :: Text -> Pat -> Pat -> Pat
+subV varNm t = \case
+  var@(VarP v) -> if v == varNm then t else var
+  x :* xs -> subV varNm t x :* subV varNm t xs
+  l := x -> subV varNm t l := subV varNm t x
+  ProdP xs -> ProdP (subV varNm t xs)
+  RecP xs -> RecP (subV varNm t xs)
+  SumP xs -> SumP (subV varNm t xs)
+  AppP t1 t2 -> AppP (subV varNm t t1) (subV varNm t t2)
+  RefP n x -> RefP (subV varNm t n) (subV varNm t x)
+  DecP a b c -> DecP (subV varNm t a) (subV varNm t b) (subV varNm t c)
+  other -> other
+
+{- Performs substitution on an entire instance (the first argument) given the
+   concrete types from a Pat (the second argument).
+   Note that ONLY PatVars which occur in the Instance *HEAD* are replaced, though they
+   are replaced in the instance superclasses as well (if they occur there).
+-}
+subst :: Rule -> Pat -> Rule
+subst cst@(C _ t :<= _) ty = mapPat (go (getSubs t ty)) cst
+  where
+    go :: [(Text, Pat)] -> Pat -> Pat
+    go subs tty =
+      let noflip p1 p2 = uncurry subV p2 p1
+       in foldl' noflip tty subs
+
+{- Given two patterns (which are hopefully structurally similar), gather a list of all substitutions
+   from the PatVars in the first argument to the concrete types (hopefully!) in the second argument
+-}
+getSubs :: Pat -> Pat -> [(Text, Pat)] -- should be a set, whatever
+getSubs (VarP s) t = [(s, t)]
+getSubs (x :* xs) (x' :* xs') = getSubs x x' <> getSubs xs xs'
+getSubs (l := t) (l' := t') = getSubs l l' <> getSubs t t'
+getSubs (ProdP xs) (ProdP xs') = getSubs xs xs'
+getSubs (RecP xs) (RecP xs') = getSubs xs xs'
+getSubs (SumP xs) (SumP xs') = getSubs xs xs'
+getSubs (AppP t1 t2) (AppP t1' t2') = getSubs t1 t1' <> getSubs t2 t2'
+getSubs (RefP n t) (RefP n' t') = getSubs n n' <> getSubs t t'
+getSubs (DecP a b c) (DecP a' b' c') = getSubs a a' <> getSubs b b' <> getSubs c c'
+getSubs _ _ = []
+
+-- NoMatch isn't fatal but OverlappingMatches is (i.e. we need to stop when we encounter it)
+data MatchError
+  = NoMatch
+  | OverlappingMatches [Rule]
+
+-- for SolveM, since we catch NoMatch
+data Overlap = Overlap Constraint [Rule]
+  deriving stock (Show, Eq)
+
+selectMatchingInstance :: Pat -> Class -> [Rule] -> Either MatchError Rule
+selectMatchingInstance p c rs = case filter matchPatAndClass rs of
```

Comment:
Oh wait I misunderstood what `RuleHead` was supposed to be below. Hm. 

I'm still not sure how this is an improvement tho. What you want is effectively `Map Constraint Rule` (I would say that `RuleHead ~ Pat`). If we're looking for the `Rule` that matches a given `Pat` for a given `Class` we can't just `lookup` using the `Class` and the `Pat`: It's an intrinsically `filter`-ey operation because we're looking for `RuleHead`s which _which are a generalization of the Pat / of which the Pat is a substitution instance_.